### ISSUE TITLE
deduplicate/fix getting the charset

### ISSUE THREAD

**Issue by ThomasWaldmann** [moinwiki/moin#920]
code like this is copy and pasted at quite some places now:

```
charset = meta['contenttype'].split('charset=')[1]
```

This works for the most simple cases, but not in general.

E.g. imagine a content-type being: `text/x.moin.wiki;charset=utf-8;format=1.9`

TODO:

- dedup this code into a helper function or Meta class method.
- generalize it so it always works

**Comment by ThomasWaldmann** [comment]
https://github.com/moinwiki/moin/blob/master/src/moin/utils/mime.py

could use `Type._parse(content_type)`.

**Comment by ThomasWaldmann** [comment]
Yeah, the parameters code might be buggy at quite some places - we can fix them all by calling the same function/method and removing the duplicated (bad) code.

> If you change the contenttype from text/x.moin.wiki;charset=utf-8 to
> charset=utf-8;text/x.moin.wiki then on save you get: [...]

The latter is not valid. it must be `type/subtype;param1=value1[;param2=value2;...]`.

> CHARSET is used in many places. We could save code if we just used CHARSET everywhere.

This is our default. That doesn't mean all content must have the default charset, so when processing content, we must read the charset from the contenttype in the metadata.

While users might rarely use something different for items interactively made with the edit form, this is important for uploads of existing content - you can then just set the charset to whatever you have in that existing document (which, for older documents often is rather iso8859-1 or some DOS/Windows charset than utf-8).

**Comment by RogerHaase** [comment]
While allowing uploads of existing content that is in a charset other than utf-8 is a desirable feature,
the user who has such content would be better served if the content were converted to utf-8 before saving. Else that user is likely to find many bugs and no support.

Repeating from contenttypes.py:

    # Charset - we support only 'utf-8'. While older encodings might work,
    # we don't have the resources to test them, and there is no real
    # benefit for the user. IMPORTANT: use only lowercase 'utf-8'!
    CHARSET = 'utf-8'

from #17 

    if a new revision gets stored into the storage, we should do some validation steps:
        * for text data, try to decode it to unicode - if it fails, do not accept it

from moin 1.9 CHANGES Version 1.3.0 (2004-12-06 ...

      Moin 1.3 uses utf-8 encoding only, there is NO SUPPORT for using a
      charset different from utf-8 with 1.3. If you try, you are on your own.

Since the code to upload and save an item in a user specified encoding does not exist, we should write code allowing the user to specify an upload encoding and convert the item to utf-8 before saving. With all the coding time saved by the simpler approach, we could offer a similar conversion on downloads.

I think we should just use CHARSET (utf-8) everywhere. There is no loss in functionality and code becomes easier to understand as we remove all the crap code extracting the charset from revision metadata.

**Comment by ThomasWaldmann** [comment]
The point is that when you upload a text file, it should be possible to get the same file back.

That would not be the case if we auto-convert to utf-8.

Also, requiring users to convert all stuff they want to upload first is also not desirable.

So, we can just store the charset and use it for decoding, problem solved.
